File: algorithms/sort/insertion_sort.js
// O(n^2) running time

const insertionSort = (nums) => {
  for (let j = 1; j < nums.length; j++) {
    let key = nums[j];
    let i = j - 1; // nums[0..i] is sorted

    while (i >= 0 && nums[i] > key) {
      nums[i+1] = nums[i];
      i = i - 1
    }
    nums[i+1] = key;
  }

  return nums;
};

console.log('result', insertionSort([2, 5, 3, 1, 6, 4]));

File: Python/algorithms/sort_and_search/insertion_sort.py
#! /usr/bin/python
"""
Time: O(N^2) time
Space: O(1) space
In-place
Stable
"""
def insertionSort(nums):
    for i in range(1, len(nums)):
        j = i
        while j > 0 and nums[j-1] > nums[j]:
            nums[j-1], nums[j] = nums[j], nums[j-1]
            j -= 1

if __name__ == "__main__":
    sample = [3, 1, 4, 5, 10, -4, 2, -5]
    print("Before sorting", sample)
    insertionSort(sample)
    print("After sorting", sample)
